src/toolkits/activity_classification/activity_classifier.cpp

```diff
@@ -464,6 +464,127 @@ gl_sframe activity_classifier::predict_per_window(gl_sframe data,
   return result;
 }
 
+gl_sframe activity_classifier::predict_topk(gl_sframe data,
+                                            std::string output_type, size_t k,
+                                            std::string output_frequency) {
+  // check valid input arguments
+  if (output_type.empty()) {
+    output_type = "probability";
+  } else if (output_type != "probability" && output_type != "rank") {
+    log_and_throw(output_type +
+                  " is not a valid option for output_type.  "
+                  "Expected one of: probability, rank");
+  }
+  if (output_frequency.empty()) {
+    output_frequency = "per_row";
+  } else if (output_frequency != "per_row" &&
+             output_frequency != "per_window") {
+    log_and_throw(output_frequency +
+                  " is not a valid option for output_frequency.  "
+                  "Expected one of: per_row, per_window");
+  }
+
+  // data inference
+  std::unique_ptr<data_iterator> data_it =
+      create_iterator(data, false, false, false);
+  gl_sframe raw_preds_per_window = perform_inference(data_it.get());
+
+  // argsort probability to get index (rank) for top k class
+  flex_list class_labels = read_state<flex_list>("classes");
+  k = std::min(k, class_labels.size());
+  auto argsort_prob = [=](const flexible_type& ft) {
```

Comment:
The code here on out is very dense, it might be helpful to add a few comments at places like this, where something not so obvious and self-explanatory is happening.

Reply:
Sure!